File: solutions/scala/all-your-base/1/src/main/scala/AllYourBase.scala
object AllYourBase {
  def rebase(
              originalBase: Int,
              originalCharacters: List[Int],
              newBase: Int
            ): Option[List[Int]] =
    if (hasInvalidInput(originalBase, originalCharacters, newBase)) {
      None
    }
    else {
      Some(this.convert(originalBase, originalCharacters, newBase))
    }

  private def hasInvalidInput(originalBase: Int, originalCharacters: scala.List[Int], newBase: Int): Boolean =
    originalBase <= 1 ||
      newBase <= 1 ||
      originalCharacters.exists(_ < 0) ||
      originalCharacters.exists(_ >= originalBase)

  private def convert(originalBase: Int, originalCharacters: List[Int], newBase: Int): List[Int] = {
    val cleanCharacters = originalCharacters.dropWhile(_ == 0)
    val base10 = this.rebaseToBase10(originalBase, cleanCharacters)
    this.returnAsTargetBase(newBase, base10)
  }

  private def returnAsTargetBase(newBase: Int, base10: Int): List[Int] =
    if (base10 == 0)
      List(0)
    else if (newBase == 10) {
      this.toIntList(base10)
    }
    else {
      this.rebaseToNewBase(newBase, base10)
    }

  @scala.annotation.tailrec
  private def rebaseToBase10(
                              originalBase: Int,
                              originalCharacters: List[Int],
                              power: Int = 0,
                              accumulated: Int = 0
                            ): Int =
    if (power == originalCharacters.size) {
      accumulated
    } else {
      this.rebaseToBase10(
        originalBase,
        originalCharacters,
        power + 1,
        addTranslatedValue(originalBase, originalCharacters, power, accumulated)
      )
    }

  private def addTranslatedValue(
                                  originalBase: Int,
                                  originalCharacters: List[Int],
                                  power: Int,
                                  accumulated: Int
                                ) = {
    val nextValuePosition = originalCharacters.size - power - 1
    val multiplier = Math.pow(originalBase, power)
    val newValue = originalCharacters(nextValuePosition) * multiplier
    accumulated + newValue.toInt
  }

  private def toIntList(base10: Int): List[Int] =
    base10.toString
      .split("")
      .map(_.toInt)
      .toList

  private def rebaseToNewBase(newBase: Int, base10: Int): List[Int] =
    this.rebaseToNewBase(newBase, base10, this.highestPowerOfBase(newBase, base10))

  @scala.annotation.tailrec
  private def rebaseToNewBase(
                               newBase: Int,
                               base10: Int,
                               highestPower: Int,
                               accumulated: List[Int] = List()
                             ): List[Int] =
    if (highestPower < 1) {
      accumulated.reverse
    }
    else {
      this.rebaseToNewBase(
        newBase,
        base10 % highestPower,
        highestPower / newBase,
        base10 / highestPower :: accumulated
      )
    }

  @scala.annotation.tailrec
  private def highestPowerOfBase(
                                  base: Int,
                                  number: Int,
                                  lastAttempted: Int = 1
                                ): Int = {
    val nextAttempt = lastAttempted * base
    if (nextAttempt > number) {
      lastAttempted
    }
    else {
      this.highestPowerOfBase(base, number, nextAttempt)
    }
  }
}

File: solutions/csharp/all-your-base/1/AllYourBase.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class AllYourBase
{
    public static int[] Rebase(int inputBase, int[] inputDigits, int outputBase) =>
        HasInvalidInput(inputBase, inputDigits, outputBase)
            ? throw new ArgumentException("Invalid Input")
            : Convert(inputBase, inputDigits.ToList(), outputBase);

    private static bool HasInvalidInput(int inputBase, int[] inputDigits, int outputBase) =>
        inputBase <= 1
            || outputBase <= 1
            || inputDigits.Any(x => x < 0)
            || inputDigits.Any(x => x >= inputBase);

    private static int[] Convert(int inputBase, List<int> inputDigits, int outputBase)
    {
        int firstCleanCharacterIndex = inputDigits.FindIndex(x => x != 0);
        if (firstCleanCharacterIndex == 1)
        {
            inputDigits.RemoveAt(0);
        }
        else if (firstCleanCharacterIndex > 1)
        {
            inputDigits.RemoveRange(0, firstCleanCharacterIndex - 1);
        }

        int base10 = RebaseToBase10(inputBase, inputDigits);
        return ReturnAsTargetBase(outputBase, base10);
    }

    private static int RebaseToBase10(int inputBase, List<int> inputDigits, int power = 0, int accumulated = 0) =>
        (power == inputDigits.Count)
            ? accumulated
            : RebaseToBase10(inputBase, inputDigits, power + 1, AddTranslatedValue(inputBase, inputDigits, power, accumulated));

    private static int AddTranslatedValue(int inputBase, List<int> inputDigits, int power, int accumulated)
    {
        int nextValuePosition = inputDigits.Count - power - 1;
        double multiplier = Math.Pow(inputBase, power);
        double newValue = inputDigits[nextValuePosition] * multiplier;
        return accumulated + (int)newValue;
    }

    private static int[] ReturnAsTargetBase(int outputBase, int base10) =>
        (base10 == 0)
            ? new int[] { 0 }
            : (outputBase == 10)
                ? ToIntList(base10)
                : RebaseToNewBase(outputBase, base10);

    private static int[] ToIntList(int base10) =>
        base10.ToString()
            .ToCharArray()
            .Select(x => int.Parse(x.ToString()))
            .ToArray();

    private static int[] RebaseToNewBase(int outputBase, int base10) =>
        RebaseToNewBase(outputBase, base10, HighestPowerOfBase(outputBase, base10), new List<int>());

    private static int[] RebaseToNewBase(int outputBase, int base10, int highestPower, List<int> accumulated) { 
        if (highestPower < 1)
        {
            return accumulated.ToArray();
        }
        else
        {
            accumulated.Add(base10 / highestPower);
            return RebaseToNewBase(outputBase, base10 % highestPower, highestPower / outputBase, accumulated);
        }
    }

    private static int HighestPowerOfBase(int outputBase, int base10, int lastAttempted = 1)
    {
        var nextAttempt = lastAttempted * outputBase;
        return (nextAttempt > base10)
            ? lastAttempted
            : HighestPowerOfBase(outputBase, base10, nextAttempt);
    }
}